helpers/location_service.py

```diff
@@ -1,166 +1,240 @@
-from PyQt6.QtCore import QThreadPool
-import websocket
+from PyQt6.QtCore import QObject, QUrl, pyqtSignal
+from PyQt6.QtWidgets import QApplication
+from PyQt6.QtWebSockets import QWebSocket
 
-from helpers import config
+from helpers import config, to_real_xy
+from parsers.maps.mapclasses import MapPoint
 
+class LocationSharingSignals(QObject):
+    textMessageReceived = pyqtSignal (str)
 
-class LocationSignals(QObject):
-    locs_recieved = pyqtSignal(dict, dict)
-    send_loc = pyqtSignal(dict)
-    death = pyqtSignal(dict)
-    config_updated = pyqtSignal()
-
-
-RUN = threading.Event()
-RUN.set()
-SIGNALS = LocationSignals()
-THREADPOOL = QThreadPool()
-_LSC = None
+class LocationSharingService(QObject):
+    character_name = None
+    group_key = None
+    host = None
+    running = False
+    websocket = None
+    zone_name = None
+    x = 0
+    y = 0
+    z = 0
 
+    def __init__(self):
+        super().__init__()
 
-def get_location_service_connection():
-    global _LSC
-    if _LSC is None:
-        _LSC = LocationServiceConnection()
-    return _LSC
+        # Connect self.config_updated to the settings signals
+        QApplication.instance()._signals["settings"].config_updated.connect(self.config_updated)
 
+        # Set self.character_name from config
+        self.character_name = config.data["sharing"]["player_name"]
 
-def start_location_service(update_func):
-    try:
-        SIGNALS.locs_recieved.disconnect()
-    except TypeError:
-        pass
-    SIGNALS.locs_recieved.connect(update_func)
-    SIGNALS.config_updated.connect(config_updated)
-    config_updated()
-    lsc = get_location_service_connection()
-    THREADPOOL.start(lsc)
+        # Set self.group_key - Handle settings changes and account for discord override
+        if config.data["sharing"]["discord_channel"]:
+            key = config.data["discord"]["url"].split("?")[0].split("/")[-1]
+            if key != "" and self.group_key != key:
+                self.group_key = key
+        else:
+            if self.group_key != config.data["sharing"]["group_key"]:
+                self.group_key = config.data["sharing"]["group_key"]
+        
+        # Set self.host from config
+        self.host = config.data["sharing"]["url"]
+
+        # Set self.zone_name from config
+        self.zone_name = config.data["maps"]["last_zone"]
+
+        # Start the service from init if sharing is enabled in the config
+        if config.data.get("sharing", {}).get("enabled", False):
+            self.start()
+
+    # Used to start the websocket service and connect signals
+    def start(self):
+        # Setup Websocket
+        self.websocket = QWebSocket()
+        self.websocket.connected.connect(self.connected)
+        self.websocket.error.connect(self.error)
+        self.websocket.disconnected.connect(self.disconnected)
+        self.websocket.textMessageReceived.connect(self.message)
+        self.websocket.open(QUrl(self.host))
```

Comment:
I need to test what happens if the user puts an empty-string in the hostname field (or in the sharing key) because users suck, though I guess worst case is they crash themselves, but then they'll go file an issue >_<